rust-runtime/inlineable/src/endpoint_lib/arn.rs

```diff
@@ -0,0 +1,138 @@
+/*
+ *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *  SPDX-License-Identifier: Apache-2.0
+ */
+
+use crate::endpoint_lib::diagnostic::DiagnosticCollector;
+use std::borrow::Cow;
+use std::error::Error;
+use std::fmt::{Display, Formatter};
+
+#[derive(Debug, Eq, PartialEq)]
+pub(crate) struct Arn<'a> {
+    partition: &'a str,
+    service: &'a str,
+    region: &'a str,
+    account_id: &'a str,
+    resource_id: Vec<&'a str>,
+}
+
+#[allow(unused)]
+impl<'a> Arn<'a> {
+    pub(crate) fn partition(&self) -> &'a str {
+        self.partition
+    }
+    pub(crate) fn service(&self) -> &'a str {
+        self.service
+    }
+    pub(crate) fn region(&self) -> &'a str {
+        self.region
+    }
+    pub(crate) fn account_id(&self) -> &'a str {
+        self.account_id
+    }
+    pub(crate) fn resource_id(&self) -> &Vec<&'a str> {
+        &self.resource_id
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub(crate) struct InvalidArn {
+    message: Cow<'static, str>,
+}
+
+impl InvalidArn {
+    fn from_static(message: &'static str) -> InvalidArn {
+        Self {
+            message: Cow::Borrowed(message),
+        }
+    }
+}
+impl Display for InvalidArn {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.message)
+    }
+}
+impl Error for InvalidArn {}
+
+impl<'a> Arn<'a> {
+    pub(crate) fn parse(arn: &'a str) -> Result<Self, InvalidArn> {
+        let mut split = arn.splitn(6, ':');
```

Comment:
There are ARNs that don't fit this pattern. Should we be more clear that this is only for parsing a specific subset of ARNs?

This isn't yet public but a comment would be nice just so future us don't get confused about that.

https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html#arns-paths

Reply:
do you have an example of an ARN you're talking about? The path at the end gets split up into the resource, and those are still 6-part ARNs

Reply:
Ah, Ok. I didn't realize the 5-colon ARNs like `arn:aws:s3:::my_corporate_bucket/Development/*` had an implicit sixth one

Reply:
there is a `::` in the middle (it's not implicit, just empty)